samples/compression/compression-custom/main.cpp

```diff
@@ -0,0 +1,261 @@
+ */
+
+#include <iostream>
+#include <fstream>
+#include <stdexcept>
+
+#include "vmf/vmf.hpp"
+
+using namespace std;
+using namespace vmf;
+
+// User-created compression algorithm should impement ICompressor interface
+// Straightforward implementation of LZW algorithm
+// Based on this: http://rosettacode.org/wiki/LZW_compression#C.2B.2B
+class MyLZWCompressor : public ICompressor
+{
+public:
+    virtual void compress(const MetaString& input, vmf_rawbuffer& output)
+    {
+        // Build the dictionary.
+        int dictSize = 256;
+        map< string, int > dictionary;
+        for (int i = 0; i < 256; i++)
+        {
+            dictionary[std::string(1, i)] = i;
+        }
+
+        vector<int> codes;
+        std::string w;
+        for (auto it = input.begin(); it != input.end(); ++it)
+        {
+            char c = *it;
+            std::string wc = w + c;
+            if (dictionary.count(wc))
+            {
+                w = wc;
+            }
+            else
+            {
+                codes.push_back(dictionary[w]); //*result++ = dictionary[w];
+                // Add wc to the dictionary.
+                dictionary[wc] = dictSize++;
+                w = std::string(1, c);
+            }
+        }
+
+        // Output the code for w.
+        if (!w.empty())
+            codes.push_back(dictionary[w]); //*result++ = dictionary[w];
+
+        //according to C++ standard, this should be correct way to convert vector to array
+        output = std::move(vmf_rawbuffer((const char*)codes.data(), codes.size()*sizeof(int)));
+    }
+
+    virtual void decompress(const vmf_rawbuffer& input, MetaString& output)
+    {
+        // Build the dictionary.
+        int dictSize = 256;
+        std::map< int, std::string > dictionary;
+        for (int i = 0; i < 256; i++)
+        {
+            dictionary[i] = std::string(1, i);
+        }
+
+        //load codes from raw buffer to vector<int>
+        size_t nCodes = input.size/sizeof(int);
+        int* bufCodes = (int*)input.data.get();
+        vector<int> codes;
+        copy_n(bufCodes, nCodes,  std::back_inserter(codes));
+
+        size_t idx = 0;
+        std::string w(1, codes[idx++]); //std::string w(1, *begin++);
+        output = w;
+        std::string entry;
+        for ( ; idx < nCodes; idx++)
+        {
+            int k = codes[idx]; //int k = *begin;
+            if (dictionary.count(k))
+            {
+                entry = dictionary[k];
+            }
+            else if (k == dictSize)
+            {
+                entry = w + w[0];
+            }
+            else
+            {
+                throw std::runtime_error("LZW decompression error: bad input index");
+            }
+
+            output += entry;
+
+            // Add w+entry[0] to the dictionary.
+            dictionary[dictSize++] = w + entry[0];
+
+            w = entry;
+        }
+    }
+
+    virtual const MetaString& getId()
```

Comment:
`MetaString` is a deprecated alias for `vmf_string` (that is likely to be changed to `vmf::String` in VMF 3.0)